Implement Input Capture Desktop Portal

3l0w opened the pull request:
#### Describe your PR, what does it fix/add?

This PR implements the input capture desktop portal with the new protocol defined in hyprwm/hyprland-protocols#8.
Fixes #259
Fixes #64

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
Nothing to declare yet.

#### Is it ready for merging, or does it need work?
It work when used with hyprwm/Hyprland#7919.

TODO List:
* [x] See if the fd multiplication problem is fixed
* [x] Calculate the invalids barriers on SetPointerBarriers
* [x] Replace the trash impl of get_line_intersection by one that follow the specs
* [x] Warp pointer when the cursor is released by the capturing program
* [x] Move most of the impl to Hyprland
* [x] Impl some kind of security around https://github.com/hyprwm/Hyprland/pull/9930, maybe move some of the impl ive done here to Hyprland
* [x] See if i limit to only one client, (for now multiple clients can connect, first arrive with his zones, first to be triggered)
* [x] Fix unable to access tty when captured

3l0w committed:
input-capture: impl portal

3l0w committed:
input-capture: impl portal

Co-authored-by: Gerben Meijer <infernix@gmail.com>
Co-authored-by: Ashley Stonham <reddrop@reddrop.net>
Co-authored-by: extra-salad <87635414+extra-salad@users.noreply.github.com>

vaxerski commented on .gitmodules:
```diff
@@ -1,6 +1,7 @@
 	path = subprojects/hyprland-protocols
-	url = https://github.com/hyprwm/hyprland-protocols
+	url = https://github.com/3l0w/hyprland-protocols
+	branch = feat/input-capture-impl
```
marking this one as a blocker until this is removed so we dont forget (don't resolve)

vaxerski commented on src/core/PortalManager.cpp:
```diff
@@ -19,13 +19,19 @@ SOutput::SOutput(SP<CCWlOutput> output_) : output(output_) {
-        transform = (wl_output_transform)transform_;
+    output->setMode([this](CCWlOutput* r, uint32_t flags, int32_t width, int32_t height, int32_t refresh) {
+        refreshRate  = refresh;
+        this->width  = width;
```
no this-> here and everywhere else

for conflicting local and member names, use `local_` e.g. `int32_t width_` and then `width = width_`

vaxerski commented on src/core/PortalManager.hpp:
```diff
@@ -32,6 +33,9 @@ struct SOutput {
     uint32_t            id          = 0;
     float               refreshRate = 60.0;
     wl_output_transform transform   = WL_OUTPUT_TRANSFORM_NORMAL;
+    uint32_t            width, height;
```
ALWAYS initialize member primitives

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+CInputCapturePortal::CInputCapturePortal(SP<CCHyprlandInputCaptureManagerV1> mgr) {
+    Debug::log(LOG, "[input-capture] initializing input capture portal");
+    m_sState.manager = mgr;
+    sessionCounter   = 0;
```
should be moved to the header

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+
+CInputCapturePortal::CInputCapturePortal(SP<CCHyprlandInputCaptureManagerV1> mgr) {
+    Debug::log(LOG, "[input-capture] initializing input capture portal");
+    m_sState.manager = mgr;
```
`CInputCapturePortal::CInputCapturePortal(SP<CCHyprlandInputCaptureManagerV1> mgr) : m_sState.manager(mgr) {`

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+    m_pObject->registerMethod(INTERFACE_NAME, "Release", "osa{sv}", "ua{sv}", [&](sdbus::MethodCall c) { onRelease(c); });
+    m_pObject->registerMethod(INTERFACE_NAME, "ConnectToEIS", "osa{sv}", "h", [&](sdbus::MethodCall c) { onConnectToEIS(c); });
+
+    m_pObject->registerProperty(INTERFACE_NAME, "SupportedCapabilities", "u", [](sdbus::PropertyGetReply& reply) { reply << (uint)(1 | 2); });
```
uint is quite ambiguous, use uint32_t

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+
+    session->session            = createDBusSession(sessionHandle);
+    session->session->onDestroy = [session, this]() {
+        if (session->status == ACTIVATED) {
```
no {} around short ifs

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+    return sessions[sessionHandle]->status != STOPPED;
+}
+
+bool get_line_intersection(double p0_x, double p0_y, double p1_x, double p1_y, double p2_x, double p2_y, double p3_x, double p3_y, double* i_x, double* i_y) {
```
no snake

3l0w replied:
Kinda wont fix, i will toss it and write a thing that is more in line with the specification of the input capture portal.

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,91 @@
+const ClientStatus ACTIVATED = 2; //Currently receiving inputs
+const ClientStatus STOPPED   = 3; //Can no longer be activated
+
+struct Barrier {
```
SBarrier

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,91 @@
+const ClientStatus CREATED   = 0; //Is ready to be activated
+const ClientStatus ENABLED   = 1; //Is ready for receiving inputs
+const ClientStatus ACTIVATED = 2; //Currently receiving inputs
+const ClientStatus STOPPED   = 3; //Can no longer be activated
```
why is this not an enum

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+void CInputCapturePortal::onAbsoluteMotion(double x, double y, double dx, double dy) {
+    for (const auto& [key, session] : sessions) {
+        int matched = session->isColliding(x, y, x - dx, y - dy);
+        if (matched != 0) {
```
no {}

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+void CInputCapturePortal::onKey(uint32_t key, bool pressed) {
+    for (const auto& [_, value] : sessions) {
```
key, value

_ is ugly, this isn't scala

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+bool CInputCapturePortal::Session::activate(double x, double y, uint32_t borderId) {
+    if (status != ENABLED) {
```
no {}

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+bool CInputCapturePortal::Session::deactivate() {
+    if (status != ACTIVATED) {
```
.

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+    double x = 0;
+    double y = 0;
+
+    if (axis) {
```
.

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,594 @@
+    int32_t x = 0;
+    int32_t y = 0;
+
+    if (axis) {
```
.

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,91 @@
+
+    void zonesChanged();
+
+    struct Session {
```
SSession

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,91 @@
+        std::string                           appid;
+        sdbus::ObjectPath                     requestHandle, sessionHandle;
+        std::string                           sessionId;
+        uint32_t                              capabilities;
```
ALWAYS initialize primitives, here and below

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+        return;
+    }
+
+    client.handle   = NULL;
```
do this in the header, also, either 0 or nullptr, NULL is C.

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+    Debug::log(LOG, "[EIS] Listening on {}", socketPath);
+
+    stop = false;
+    std::thread thread(&EmulatedInputServer::listen, this);
```
this is bad, get the fd and incorporate it into the event loop poll thread in the main portalmgr.cpp

3l0w replied:
This was not straightforward to impl, because i am creating one socket per session. I implemented something with a vector and a map inside the portalmgr (`addFdToEventLoop`, `removeFdFromEventLoop`)

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+int EmulatedInputServer::onEvent(eis_event* e) {
+    eis_client* client;
+    eis_seat*   seat;
+    eis_device* device;
```
= nullptr all

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+                return 0;
+            }
+
+            this->client.handle = client;
```
no this->

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+                Debug::log(WARN, "[EIS] Unknown device to close");
+            }
+            break;
+        case EIS_EVENT_FRAME: Debug::log(LOG, "[EIS] Got event EIS_EVENT_FRAME"); break;
```
shouldn't these do something more than just logging? Also, shouldn't these be dropped to TRACE? Kinda verbose.

3l0w replied:
I forgot to remove them.

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+}
+
+void EmulatedInputServer::ensurePointer(eis_event* event) {
+    if (client.pointer != nullptr)
```
!= nullptr redundant

vaxerski commented on src/shared/Eis.cpp:
```diff
@@ -0,0 +1,286 @@
+    if (client.pointer != nullptr)
+        return;
+
+    struct eis_device* pointer = eis_seat_new_device(client.seat);
```
C-style structs, drop the struct

vaxerski commented on src/shared/Eis.hpp:
```diff
@@ -0,0 +1,48 @@
+#include <string>
+
+struct EisClient {
+    struct eis_client* handle;
```
no C-style structs + initialize

vaxerski reviewed (commented):
a bit of a lengthy review, mostly style and guidelines. Many comments apply to the entirety of the PR

vaxerski commented:
tag me for review when ready

3l0w commented:
I am aware is due to the sdbus version bumb, it should be good now

Azodox commented:
I am having this:

```
In file included from /home/luke/xdg-desktop-portal-hyprland/src/core/PortalManager.hpp:11,
                 from /home/luke/xdg-desktop-portal-hyprland/src/core/PortalManager.cpp:1:
/home/luke/xdg-desktop-portal-hyprland/src/core/../portals/InputCapture.hpp:2:10: fatal error: dbusDefines.hpp: No such file or directory
    2 | #include "dbusDefines.hpp"
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/xdg-desktop-portal-hyprland.dir/build.make:125: CMakeFiles/xdg-desktop-portal-hyprland.dir/src/core/PortalManager.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:102: CMakeFiles/xdg-desktop-portal-hyprland.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
-- Install configuration: ""
-- Installing: /usr/bin/hyprland-share-picker
-- Up-to-date: /usr/bin/hyprland-share-picker
CMake Error at build/cmake_install.cmake:81 (file):
  file INSTALL cannot find
  "/home/luke/xdg-desktop-portal-hyprland/build/xdg-desktop-portal-hyprland":
  No such file or directory.
```

littleblack111 commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,100 @@
+#pragma once
+#include "dbusDefines.hpp"
```
```suggestion
#include "../dbusDefines.hpp"
```

littleblack111 commented:
https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/268/files#r1817726512

should fix it

3l0w commented:
I pushed the fix

littleblack111 commented:
i gave feedback in https://github.com/hyprwm/Hyprland/pull/7919

gptlang commented:
```
Hyprland 0.44.0 built from branch  at commit 189f6152f1354ae936cafffa7b7dd62af8dd5bd4  (input-capture: fix rebase issues).
```

Deskflow:
```
[2024-11-05T00:39:23] ERROR: failed to initialize input capture session, quitting: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.InputCapture” on object at path /org/freedesktop/portal/desktop
```

Is this currently working?

littleblack111 commented:
u need [xdg-desktop-portal](https://github.com/3l0w/xdg-desktop-portal-hyprland/tree/feat/input-capture-impl) too

littleblack111 commented:
it works for me. restart ur xdg-desktop-portals after installation

3l0w commented:
Also make sure that the hyprland.portal file has been copied over `/usr/share/xdg-desktop-portal/portals`

littleblack111 commented:
also, as an update, still consistently have my modifiers always held down on the client and have to restart input leap to fix

gptlang commented:
It seems to work fine now as the server. `org.freedesktop.portal.RemoteDesktop` is still needed for the client though.

littleblack111 commented:
also @3l0w, input capture should not work in ext-session-lock

littleblack111 commented:
> It seems to work fine now as the server. `org.freedesktop.portal.RemoteDesktop` is still needed for the client though.

dont know if this will work: https://github.com/waycrate/xdg-desktop-portal-luminous since hyprland is no longer wlroots. but it does have `org.freedesktop.portal.RemoteDesktop`

mewoocat commented:
Any updates on the status of this PR?

3l0w commented:
> > Any updates on the status of this PR?
> 
> have you tested it? I believe the following problems have to be solved before merge:
> 
>     1. instability, still have to restart it few times a day
> 
>     2. not sure why, i did not bind anything on top of my display, but it will attempt to capture my mouse when i put it on the edge of the top of my display(doesnt happen to right or bottom, i binded left), behavoir can be seen as the mouse disappearing(captured) then reappearing shortly after(since it's input-leap have nothing to put it to or just refused or sth)
> 
>     3. prob need an configurable option to release the current keys/modi when switching screens
> 
>     4. isLock() delay meaning if your mouse is captured by an application on ur server, and around the same time, you put ur cursor to the edge, it can be going thru(so either fix isLock()(which idk how to do), or [fix: forcerelease when locked && captured 3l0w/Hyprland#3](https://github.com/3l0w/Hyprland/pull/3) which have on going discussion on how should the release be done
> 
> 
> EDIT: did not relize there is new commits here in xdph too, will test them out

Top capturing don't work ? do you have a screen binded on top ? Even disabled ?

littleblack111 commented:
nope. no screen binded on top. only when enabled

3l0w commented:
For the ZonesChanged to work properly https://github.com/flatpak/xdg-desktop-portal/pull/1605 and https://github.com/input-leap/input-leap/pull/2166 are needed.
Also if you have your left monitor position that is not at 0 0 you probably need this also: https://github.com/input-leap/input-leap/pull/2167 (this fix top and/or left edge that don't trigger)

gestate-8 commented:
Does this also work with libei? It works with Xwayland on my device, but with libei it didn't work with error 'No such interface "org.freedesktop.portal.RemoteDesktop".'

3l0w commented:
Libei is just the transport between xdph and the client app (eg. input-leap).
This PR just aim to implement the Input Capture portal (the portal used to share the mouse input to a program).
Remote Desktop is another portal which aim to receive and execute input from a program.
The implement of Remote Desktop is currently in #308.

lucianolacurcia commented:
+1 I am still using i3wm because I cant use deskflow in hyprland yet as it requires input-capture implementation

3l0w commented:
> +1 I am still using i3wm because I cant use deskflow in hyprland yet as it requires input-capture implementation

You can install my fork of Hyprland & XDPH to make it work

ackstorm23 commented:
> +1 I am still using i3wm because I cant use deskflow in hyprland yet as it requires input-capture implementation

I use Waynergy as a client in hyprland. I have not tried it with deskflow but it's working with my synergy host.

Vaisakhkm2625 commented:
> I think we are still weeks away from merging. Here are my todo list:
> 

Wait, only weeks away... ? I have been waiting for years for this and i am too dumb even to think about doing myself... 
and was thinking that day would never come.. You are doing gods work.. Thank you..

CaptaiNiveau commented:
Yeah, I'm also really looking forward to this! I'll see if I can try it this week, would be awesome to use my laptop to control my PC when it's connected to the TV ^^

3l0w commented:
You can already use it, it will mostly work, if you have problm to setup it feel free to poke me on discord @3l0w

SMQuazi commented:
Just adding information for anyone it may help. Installing it on CachyOS (arch) was mostly smooth but got stuck for a bit on figuring out that I had to copy `hyprland-input-capture-v1.xml` from the `hyprland-protocols`  repo to `/usr/share/hyprland-protocols/protocols/`, despite cloning recursively.  

During install it did complain that sdbus didn't have `createSessionBusConnection`. I'm not entirely sure what fixed that issue as I have the 
 current `sdbus-cpp` package, which is what I tried initially.

Once the installation was done, Input-Leap (synergy based) worked pretty much flawlessly. The only minor hiccup is that when I use the `exit` command from hyprland and log back in, it complains about the address already being bound. I've tried to shut all related processes down and restarting input-leap which allows it to connect and running seemingly fine, but I can't seem to actually move my mouse to a different "screen"/device. So I ended up replacing the `exit` keybind with `pkill -u $USER`. Probably not the most graceful solution, but I've had no issues so far.

Thanks for all the great work. This is very exciting and personally allows hyprland to be the only DE I need to have installed.

CaptaiNiveau commented:
I also ran into that issue with sdbus - it was because I'm stupid and didn't notice I'm on the main branch of the PR, not the feature branch. When you switch to the feature branch, it should be fine.

gac3k commented:
@3l0w hey, friendly bump here. Please just let us know if you want to finish it, otherwise I guess we can take it over so that it can finally be merged.

joeesteves commented:
> Gentle bump on this. I rebased this and compiled the portal, and input capture is working really well. I'm using deskflow with no issues.

Same for me, using I was able to run deskflow 👌🏼

Kbz-8 commented:
every time I try to build I get this error
```
[  2%] Generating /home/kbz8/Downloads/xdg-desktop-portal-hyprland/protocols/hyprland-input-capture-v1.cpp, /home/kbz8/Downloads/xdg-desktop-portal-hyprland/protocols/hyprland-input-capture-v1.hpp
Couldn't load proto
make[2]: *** [CMakeFiles/xdg-desktop-portal-hyprland.dir/build.make:116: /home/kbz8/Downloads/xdg-desktop-portal-hyprland/protocols/hyprland-input-capture-v1.cpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:110: CMakeFiles/xdg-desktop-portal-hyprland.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I tried [this](https://github.com/hyprwm/xdg-desktop-portal-hyprland/issues/327) but it did not work. Any idea ?

Btw cmake finds these versions of the dependencies
```
-- Checking for module 'hyprutils>=0.2.6'
--   Found hyprutils, version 0.8.3
-- Checking for modules 'wayland-client;wayland-protocols;libpipewire-0.3>=1.1.82;libspa-0.2;libdrm;libeis-1.0;uuid;gbm;hyprlang>=0.2.0;hyprutils>=0.2.6;hyprwayland-scanner>=0.4.2'
--   Found wayland-client, version 1.24.0
--   Found wayland-protocols, version 1.45
--   Found libpipewire-0.3, version 1.4.7
--   Found libspa-0.2, version 0.2
--   Found libdrm, version 2.4.125
--   Found libeis-1.0, version 1.5.0
--   Found uuid, version 2.41.1
--   Found gbm, version 25.2.1-arch1.4
--   Found hyprlang, version 0.6.3
--   Found hyprutils, version 0.8.3
--   Found hyprwayland-scanner, version 0.4.5
```

Kbz-8 commented:
> Saaoa6&9 7

wat ?

3l0w commented:
Did you pulled the git submodules ?
```bash
git submodule init
git submodule update
```

R0wDrunner commented:
> > > Did you pulled the git submodules ?
> > > ```shell
> > > git submodule init
> > > git submodule update
> > > ```
> > 
> > 
> > I cloned recursively. Just tried this and nothing changed
> 
> Did you try doing:
> 
> ```
> git clone https://github.com/3l0w/xdg-desktop-portal-hyprland.git --recurse-submodules
> ```

Would you be able to do a writeup on how to apply this, e.g. for using Deskflow properly - if this PR works for you?

stauersbol commented:
> > > > Did you pulled the git submodules ?
> > > > ```shell
> > > > git submodule init
> > > > git submodule update
> > > > ```
> > > 
> > > 
> > > I cloned recursively. Just tried this and nothing changed
> > 
> > 
> > Did you try doing:
> > ```
> > git clone https://github.com/3l0w/xdg-desktop-portal-hyprland.git --recurse-submodules
> > ```
> 
> Would you be able to do a writeup on how to apply this, e.g. for using Deskflow properly - if this PR works for you?

I did all of this last week and it worked like a charm. But sure, I can definitely try.

stauersbol commented:
> > > > > > Did you pulled the git submodules ?
> > > > > > ```shell
> > > > > > git submodule init
> > > > > > git submodule update
> > > > > > ```
> > > > > 
> > > > > 
> > > > > I cloned recursively. Just tried this and nothing changed
> > > > 
> > > > 
> > > > Did you try doing:
> > > > ```
> > > > git clone https://github.com/3l0w/xdg-desktop-portal-hyprland.git --recurse-submodules
> > > > ```
> > > 
> > > 
> > > Would you be able to do a writeup on how to apply this, e.g. for using Deskflow properly - if this PR works for you?
> > 
> > 
> > I did all of this last week and it worked like a charm. But sure, I can definitely try.
> 
> Haha I tried it too, but maybe I don't have enough experience with Linux or something, it definitely didn't work for me

Try this:
https://gist.github.com/stauersbol/235884919606ea3584c0db48d9c94129

Made a gist as to avoid sending a wall of text here.

Kbz-8 commented:
> > > Did you pulled the git submodules ?
> > > ```shell
> > > git submodule init
> > > git submodule update
> > > ```
> > 
> > 
> > I cloned recursively. Just tried this and nothing changed
> 
> Did you try doing:
> 
> ```
> git clone https://github.com/3l0w/xdg-desktop-portal-hyprland.git --recurse-submodules
> ```

well it works but I have compilations errors, I'll try older commits, thank you

stauersbol commented:
> > > > Did you pulled the git submodules ?
> > > > ```shell
> > > > git submodule init
> > > > git submodule update
> > > > ```
> > > 
> > > 
> > > I cloned recursively. Just tried this and nothing changed
> > 
> > 
> > Did you try doing:
> > ```
> > git clone https://github.com/3l0w/xdg-desktop-portal-hyprland.git --recurse-submodules
> > ```
> 
> ~well it works but I have compilations errors, I'll try older commits, thank you~ nope, I forgot to checkout to the feature branch, I still have the same issue

then try my guide that i sent earlier and see if you get it to work

R0wDrunner commented:
> every time I try to build I get this error
> 
> ```
> [  2%] Generating /home/kbz8/Downloads/xdg-desktop-portal-hyprland/protocols/hyprland-input-capture-v1.cpp, /home/kbz8/Downloads/xdg-desktop-portal-hyprland/protocols/hyprland-input-capture-v1.hpp
> Couldn't load proto
> make[2]: *** [CMakeFiles/xdg-desktop-portal-hyprland.dir/build.make:116: /home/kbz8/Downloads/xdg-desktop-portal-hyprland/protocols/hyprland-input-capture-v1.cpp] Error 1
> make[1]: *** [CMakeFiles/Makefile2:110: CMakeFiles/xdg-desktop-portal-hyprland.dir/all] Error 2
> make: *** [Makefile:136: all] Error 2
> ```
> 
> I tried [this](https://github.com/hyprwm/xdg-desktop-portal-hyprland/issues/327) but it did not work. Any idea ?
> 
> Btw cmake finds these versions of the dependencies
> 
> ```
> -- Checking for module 'hyprutils>=0.2.6'
> --   Found hyprutils, version 0.8.3
> -- Checking for modules 'wayland-client;wayland-protocols;libpipewire-0.3>=1.1.82;libspa-0.2;libdrm;libeis-1.0;uuid;gbm;hyprlang>=0.2.0;hyprutils>=0.2.6;hyprwayland-scanner>=0.4.2'
> --   Found wayland-client, version 1.24.0
> --   Found wayland-protocols, version 1.45
> --   Found libpipewire-0.3, version 1.4.7
> --   Found libspa-0.2, version 0.2
> --   Found libdrm, version 2.4.125
> --   Found libeis-1.0, version 1.5.0
> --   Found uuid, version 2.41.1
> --   Found gbm, version 25.2.1-arch1.4
> --   Found hyprlang, version 0.6.3
> --   Found hyprutils, version 0.8.3
> --   Found hyprwayland-scanner, version 0.4.5
> ```

Try to check the gist of @stauersbol https://gist.github.com/stauersbol/235884919606ea3584c0db48d9c94129
Specifically check my comment, because I ran into the same issue as you did

KeesTucker commented:
Awesome work! Followed the gist instructions and after reading the comments had minimal issues (am running Arch). Input Leap just worked after a reboot! Thank you :)

frob commented:
I followed the gist and installed this branch. It mostly works for Deskflow as server but not as client. Not sure if this is supposed to work as client because that is the Remote Desktop feature and not the Input Capture feature.

I say mostly because it only works if the client is configured to be to the left or top of the server. My guess is it isn't reading the width/height of the desktop correctly to push off to the right or down.

3l0w commented:
Yeah RemoteDesktop is out of the scope of this PR.
There is a PR open for that based on mine, but i forgot which one.
About the key repeat, i will look into it, but idk how do this work on the compositor side

3l0w commented:
I don't know how you installed xdph, i assume its a systemd type thing so maybe try:
```
journalctl --user -u xdg-desktop-portal-hyprland
```

frob commented:
My setup is using the Omarchy script and then running through the [gist](https://gist.github.com/stauersbol/235884919606ea3584c0db48d9c94129) to get this installed. I will post the logs soon. I am not logged into github on that computer.

frob commented:
Running `journalctl --user -u xdg-desktop-portal-hyprland` said there are no entries.

stauersbol commented:
> Running `journalctl --user -u xdg-desktop-portal-hyprland` said there are no entries.

you should be able to run:
```
systemctl --user xdg-desktop-portal-hyprland -n 200
```

Apparently journal can't access the service for some reason (I ssh'd to my hyprland setup to check the usual stuff where journal stores files etc and couldn't find it) I'm no expert on this.

frob commented:
Another update, I am running this on a desktop with 1440x2560 native display and no scaling. It is working as intended. I don't know if there is an issue with the other machine being a macbook pro with retina and scaling or if it is the internal trackpad that is the issue. If I can get around to testing it with a wired mouse and/or scaling set to 1, I will let you know the results.

3l0w commented:
> Another update, I am running this on a desktop with 1440x2560 native display and no scaling. It is working as intended. I don't know if there is an issue with the other machine being a macbook pro with retina and scaling or if it is the internal trackpad that is the issue. If I can get around to testing it with a wired mouse and/or scaling set to 1, I will let you know the results.

Maybe an issue on deskflow side, i don't manage at all sending inputs through network

frob commented:
> > Another update, I am running this on a desktop with 1440x2560 native display and no scaling. It is working as intended. I don't know if there is an issue with the other machine being a macbook pro with retina and scaling or if it is the internal trackpad that is the issue. If I can get around to testing it with a wired mouse and/or scaling set to 1, I will let you know the results.
> 
> Maybe an issue on deskflow side, i don't manage at all sending inputs through network

At his point in testing, I expect this to be a bug in the hyprland monitor setup and how it relates to a quirk in the old Intel 13 inch macbook pro I was testing on initially.

As for the keyhold down issue, I hope to install the update this weekend. I am also noticing this. I don't think this is an issue with deskflow as it worked when I was running deskflow on another os.

3l0w commented:
But yeah #259 i will add it

sithlord48 commented:
> This MR is not for RemoteDesktop, for that see https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/308

Yeah sorry i had intended to leave the comment on that PR 

> But yeah https://github.com/hyprwm/xdg-desktop-portal-hyprland/issues/259 i will add it

FYI You seam to have unlinked closing of #64 when you added it

3l0w commented:
For people having issue due to repeated input not working, i suggest u to test on gnome & kde the issue, because i "think" its an issue on my side, since i don't have to manage that, i have only have to send things like "user pressed this key", "user released that key"

zhurai commented:
> For people having issue due to repeated input not working, i suggest u to test on gnome & kde the issue, because i "think" its an issue on my side, since i don't have to manage that, i have only have to send things like "user pressed this key", "user released that key"

@3l0w I haven't tried the new push yet.

But, I reported the repeated input (holding down a key) issue as it did work properly in my old setups (Wayland KDE to Mac clients at least). However, after switching to this setup to test (because an unrelated thing that I ran into in KDE), I noticed that it didn't work the same.

3l0w commented:
Ive checked the code again, i sent the key the same way as it would to an application.
So i kinda don't know how to fix that issue 
<img width="1137" height="934" alt="image" src="https://github.com/user-attachments/assets/8eac7391-0949-48dd-b0e9-dfb5c10488e0" />

3l0w commented:
> Perhaps there is a modifier or status that isn't getting sent over. I noticed the other day that the numlock/capslock isn't being sent over either. I had to enable numlock on the remote machine even though it was already enabled on the keyboard and the host machine respected it.

When input is captured, modifiers are too and sent to the other computer, it work on my side with input-leap linux to linux.

You can try to apply this patch on Hyprland (`git apply {file}`), and tell me if you have better results
```patch
diff --git a/src/managers/input/InputManager.cpp b/src/managers/input/InputManager.cpp
index dd4066da..3700abd8 100644
--- a/src/managers/input/InputManager.cpp
+++ b/src/managers/input/InputManager.cpp
@@ -1513,10 +1513,6 @@ void CInputManager::onKeyboardMod(SP<IKeyboard> pKeyboard) {
     MODS.depressed     = ALLMODS;
     m_lastMods         = MODS.depressed;
 
-    PROTO::inputCapture->modifiers(MODS.depressed, MODS.latched, MODS.locked, MODS.group);
-
-    if (PROTO::inputCapture->isCaptured())
-        return;
     const auto IME = m_relay.m_inputMethod.lock();
 
     if (IME && IME->hasGrab() && !DISALLOWACTION) {
```

3l0w commented:
@vaxerski PR is ready for review

Felos9001 commented:
waiting for this PR to merge as I also have the issue of not being able to use input capture on Hyprland.

syorito-hatsuki commented:
Waiting to use Input Leap and feel again power of 2 PC and 5 monitor setup xD

SMQuazi commented:
Something weird started happening recently. I'm not sure if it's input-leap or this implementation , but whenever I start the server, there's a portion of my screen that I can no longer move to. It's roughly the size of my side monitors

My hyprland setup
```
monitor=DP-1, 2560x1440@120, -1440x0, 1, transform, 1
monitor=HDMI-A-1, 3840x2160@120, 0x0, 1, bitdepth, 10, cm, hdr, sdrbrightness, 1.2, sdrsaturation, 1, vrr, 3
monitor=DP-2, 2560x1440@120, 3840x0, 1, transform, 3
```
the box I can't enter is about 1120 x 1440

fwiw, this doesn't happen in KDE with input-leap so I'm inclined to believe it has something to do with the implementation, which I've used for a while now. Only since the last update that broke my hyprland install has this been happening.

frob commented:
I recently updated to the latest on this branch. All in all this has been very stable.

RE the numlock issue I was having. Pretty sure this is an issue with my machine. I am running Omarchy and this setup has a utility that messes with Capslock/Numlock. Either way, it remains consistent. Every time I move from my server to a client or back I have to turn on Numlock.

The repeat key still doesn't seem to carry over from server to client. This is, by far, the most annoying issue for me.

Something new that I noticed is that keyboard events are no longer being captured if the server is locked. This used to work but it no longer does. I don't know if this was expected behavior or if this is something to do with my setup. Mouse events continue to propagate, move and click.

Kaysera commented:
I also updated to the last version, and it works almost great, but I'm finding that if the server goes to sleep, then when it resumes libei seems to be locked and I have to restart. Using input-leap, I get the error `[2025-11-13T09:55:56] NOTE: Failed to set up ei: Operation not permitted`, but I don't know how to find something more descriptive. 

I'm not sure if it has something to do with this https://github.com/input-leap/input-leap/issues/2325, but it does not ask for permissions, there is no popup.

3l0w commented:
> I also updated to the last version, and it works almost great, but I'm finding that if the server goes to sleep, then when it resumes libei seems to be locked and I have to restart. Using input-leap, I get the error `[2025-11-13T09:55:56] NOTE: Failed to set up ei: Operation not permitted`, but I don't know how to find something more descriptive.
> 
> I'm not sure if it has something to do with this [input-leap/input-leap#2325](https://github.com/input-leap/input-leap/issues/2325), but it does not ask for permissions, there is no popup.

Is this consistent ? I will try on my side
And no its not related to that issue you linked, as i said in https://github.com/hyprwm/Hyprland/pull/7919#issuecomment-3345909269, you need to enable permission enforcement.

Kaysera commented:
I have been checking this today, and it was a permission problem. Now with them enforced the pop-up appears and works as it should, even after waking from sleep. However, I have been finding some more bugs using the latest version of input-leap, 3.0.3:

- If you [disable a monitor](https://wiki.hypr.land/Configuring/Monitors/#disabling-a-monitor), it shows a warning disconnecting from EIS and the permissions pop-up will not appear again (and of course input-leap does not work anymore), only reappearing after a reboot. 
- The permissions button "apply and remember" does not work as it should. It allows the app but upon reboot it does not remember the permissions
- On that subject, I have not been able to allow them through config either (https://wiki.hypr.land/Configuring/Permissions/). I have tried allowing xdg-desktop-portal-hyprland for keyboard and screencopy, as the popup shows that the one accessing the input is  xdg-desktop-portal-hyprland. Should I be allowing input-leap instead? And which one is the correct permission?

If you need me to log anything or give more details, I'm totally available. I wrote down what I thought would be relevant, but if anything else is needed, I'll try to help as much as I can 😁

3l0w commented:
> I have been checking this today, and it was a permission problem. Now with them enforced the pop-up appears and works as it should, even after waking from sleep. However, I have been finding some more bugs using the latest version of input-leap, 3.0.3:
> 
>     * If you [disable a monitor](https://wiki.hypr.land/Configuring/Monitors/#disabling-a-monitor), it shows a warning disconnecting from EIS and the permissions pop-up will not appear again (and of course input-leap does not work anymore), only reappearing after a reboot.
> 
>     * The permissions button "apply and remember" does not work as it should. It allows the app but upon reboot it does not remember the permissions
> 
>     * On that subject, I have not been able to allow them through config either (https://wiki.hypr.land/Configuring/Permissions/). I have tried allowing xdg-desktop-portal-hyprland for keyboard and screencopy, as the popup shows that the one accessing the input is  xdg-desktop-portal-hyprland. Should I be allowing input-leap instead? And which one is the correct permission?
> 
> 
> If you need me to log anything or give more details, I'm totally available. I wrote down what I thought would be relevant, but if anything else is needed, I'll try to help as much as I can 😁

This MR doesn't aim to fix permission manager issues, the popup not appearing until reboot is a normal behavior.
You should allow xdg-desktop-portal-hyprland with "input-capture", but i forget to add it so there is a new commit with the fix on the Hyprland repo
And the monitor disabled who broke everything is quite strange, since i use it everyday

frob commented:
I updated to the latest a couple of days ago and now deskflow doesn't work. I have uninstalled it and I am working through getting it working again. It looks like stuff works until I get to the barrier between server and client and the input disappears. It wont come back to either machine. I have to open a new TTY and switch back to Hypr to get the input back. Odd thing seeing as how the only input that worked was CTL + ALT + F2 to get to the TTY.

sithlord48 commented:
deskflow ignores the ctrl+alt+f# shortcuts for this reason, that is in case you wanted to switch ttys

ZeroSegFault commented:
My experience is still rock stable I have not updated since my last post with the various patches to fix other users reported issues as I was not experiencing them.

I do have the issue when scaling is greater than 1 I can't leave the screen to the right. I worked around this by increasing font sizes in my various apps individually for now.

I am hesitant to try the latest commit but if it helps I can.

infernix commented:
So if deskflow works with scaling on GNOME and KDE, then I assume the problem is with the input capture portal not taking into account scaling. I see this in the logs:

```
[LOG] [input-capture] New GetZones request:
[LOG] [input-capture]  | /org/freedesktop/portal/desktop/request/1_662/portal1573945700
[LOG] [input-capture]  | /org/freedesktop/portal/desktop/session/1_662/portal2060071402
[LOG] [input-capture]  | appid: 
[LOG] [input-capture]  | w: 2560 h: 1600 x: -1280 y: 1280
[LOG] [input-capture]  | w: 7680 h: 4320 x: 0 y: 0
...
[LOG] [input-capture] New SetPointerBarriers request:
[LOG] [input-capture]  | /org/freedesktop/portal/desktop/request/1_662/portal1318835748
[LOG] [input-capture]  | /org/freedesktop/portal/desktop/session/1_662/portal2060071402
[LOG] [input-capture]  | appid: 
[LOG] [input-capture]  | zoneSet: 2
[LOG] [input-capture]  | barrier: 1, [-1280, 1280] [1279, 1280] valid: true
[LOG] [input-capture]  | barrier: 1, [0, 0] [7679, 0] valid: true
...
```
I'm guessing here, but if SetPointerBariers uses the unscaled geometry, and scaling = 2 on both screens, then the laptop screen is really 1280x800 left of the main screen at 3840x2160. And so this is where i suppose the barrier is drawn too far because barrier 1 goes all the way from -1280 to 1280, where it should be from -1280 to 0.

frob commented:
> > My experience is still rock stable I have not updated since my last post with the various patches to fix other users reported issues as I was not experiencing them.
> > I do have the issue when scaling is greater than 1 I can't leave the screen to the right. I worked around this by increasing font sizes in my various apps individually for now. 
> 
> I updated to the latest commit. After posting this last message. Been stable and great! Looking forward to testing scaling :)

Can you also test with deskflow setup with the clients on different sides? I wonder if the issue I had originally with my old mac book pro was due to scaling issues. Specifically, it didn't work for me, before, when I had my client on the right or below the server.

ZeroSegFault commented:
> That is the issue I was having as well.

Yeah @3l0w  mentioned he missed the scaling implementation. So once that is out ill be one of the first to test it :)

infernix committed:
input-capture: fix mixed-DPI barrier handling and id mapping (#3)

Switch zone/barrier handling to logical coordinates and introduce per-session internal barrier id remapping to prevent invalid/misrouted activation events.

zacpod commented:
This isn't working for me with a 2x2 monitor setup.  If I leave the config alone and let it be 1x4 then it works fine.  But as soon as I switch config to match the physical setup, Synergy can no longer apply the barrier.

zacpod commented:
> Hi @3l0w any word on when you might have time to rebase? It has been really useful for me and I am so appreciative to have Deskflow working in Hyprland with this PR but the versions are starting to fall behind. I only wish i could be more useful than a tester but this code has been of genuine use to me and made my life easier.

I'm in the same boat.  This code is what is allowing me to run Hyprland instead of KDE.  I don't mind being a little behind, though.    It'll be a glorious day when this finally gets rolled in to the main codebase, though!

zacpod commented:
> Sorry, for the time this is taking, it also drive me nuts, i am waiting for [hyprwm/hyprpaper@9df6ba9](https://github.com/hyprwm/hyprpaper/commit/9df6ba9e9f5c330fe673ba852b51b487353d658d) to be released. i don't use the whole -git stack on my computer (only hypr and xdph), so i cannot install hyprpaper-git without installing every hypr*-git package. And without this commit i cannot build hypr from main. I will probably try this W-E to do it

No stress, dude - your work is VERY appreciated!

3l0w committed:
input-capture: little log in get zones

zacpod commented:
I'm having an issue with games holding the mouse cursor on this 0.54.0 version.  Mouse (on a 2x2 multi-monitor setup) won't stay on the running game's monitor.  I /think/ this isn't an issue with this patch, but something with base 0.54.0 but it has made me have to flip back to KDE for now.  I'm hoping it's fixed in 0.54.1.

prurigro commented:
I managed to get this working, but only seem to be able to connect to clients to the left or above the server. Both right and below hit the sides without being able to extend beyond them.

prurigro commented:
It looks like the issue might have to do with scaling. I have a 1080p monitor next to a 4k monitor. If I set the 4k monitor to a scale factor of 1, I can move from the bottom of that onto the client below it. I can't move from the bottom of the 1080p monitor, presumably because it's not positioned against the bottom. With a scale factor of 2 on the 4k monitor I can't move beyond the bottom of either display.

EDIT: Yeah, with scaling set to 1 and the vertical offset of the 1080p monitor set to 1080 I can move down from both. It seems like maybe the scale factor isn't taken into account when determining the point to jump off from. Interestingly, it doesn't seem to help when I set the scale factor to 2 and then move both monitors down by 1080.

EDIT 2: It seems like patches for scale factor were already added above... Is it possible that they didn't account for mixed DPI?

prurigro commented:
I'm using swayidle to turn off my displays with `hyprctl dispatch dpms off` and after resuming, Deskflow wasn't able to interface with ei and xdg-desktop-portal-hyprland started using 100% CPU. Killing xdg-desktop-portal-hyprland made everything work correctly again.

I tested a second time by manually running dpms off and back on, but it didn't happen. I believe I triggered this last night while changing the display settings though, so maybe there's some sort of race condition happening?

EDIT: It looks like something similar to this was reported last week in the sister-PR for hyprland

EDIT 2: I meant to say hypridle (not swayidle), only noticed it thanks to ZeroSegFault

ZeroSegFault commented:
> I'm using swayidle to turn off my displays with `hyprctl dispatch dpms off` and after resuming, Deskflow wasn't able to interface with ei and xdg-desktop-portal-hyprland started using 100% CPU. Killing xdg-desktop-portal-hyprland made everything work correctly again.
> 
> 
> 
> I tested a second time by manually running dpms off and back on, but it didn't happen. I believe I triggered this last night while changing the display settings though, so maybe there's some sort of race condition happening?
> 
> 
> 
> EDIT: It looks like something similar to this was reported last week in the sister-PR for hyprland

I am doing the same thing but with hypridle and not experienced this.

prurigro commented:
@ZeroSegFault how many displays are you running?

ZeroSegFault commented:
> @ZeroSegFault how many displays are you running?

3 displays Hyprland in the middle and a Mac on each side

prurigro commented:
Hmm, it just happened again- most of my applications were gone after resuming too...

@ZeroSegFault Does your server have a single display? Just trying to figure out what might be different.

zacpod commented:
> Hmm, it just happened again- most of my applications were gone after resuming too...
> 
> @ZeroSegFault Does your server have a single display? Just trying to figure out what might be different.

I've had similar things - usually bringing up a special workspace and then sending it away "resets" the display to it's normal workspace and my apps appear.

ZeroSegFault commented:
> Hmm, it just happened again- most of my applications were gone after resuming too...
> 
> 
> 
> @ZeroSegFault Does your server have a single display? Just trying to figure out what might be different.

Most of the time yeah the server has a single display. But I have also had it working with 2 screens on the server

infernix commented:
> Hmm, it just happened again- most of my applications were gone after resuming too...

just a note here, i had applications crash on suspend/screen off with hypridle/hyprlock on 0.54.2 and upgrading to 0.54.3 fixed that specifically.

prurigro commented:
I just triggered the 100% CPU xdg-desktop-portal-hyprland issue without losing any applications when I changed monitor config, so I think they might be separate issues.

EDIT: I'll sometimes see the same issue with wireplumber on gnome after waking from sleep if deskflow is enabled, so it's possible the issue stems from deskflow and not these patches.

prurigro commented:
So I did a bit more dual-scale-factor testing. Moving the cursor to the bottom of the screen triggers the permission popup, and I noticed that while Deskflow is enabled, moving the mouse close to the bottom causes a small amount of jitter in the cursor on both the 1080p monitor and the 4k one with scaling set to 2 (jitter that isn't there without Deskflow running).

To me this would suggest that the barriers are in the correct place and are possibly even attempting to trigger, but Deskflow doesn't seem to notice as far as I can tell based on its logs.

ZeroSegFault commented:
@3l0w  

I’ve been using this input-capture work for Deskflow for a while now and had a go at getting it working again on current Hyprland.

  I’ve got both sides forward-ported and running locally now, and Deskflow is working for me.

  Branches are here:

  - Hyprland: https://github.com/ZeroSegFault/Hyprland/tree/zerosegfault/input-capture-hyprland
  - xdg-desktop-portal-hyprland: https://github.com/ZeroSegFault/xdg-desktop-portal-hyprland/tree/input-capture-portal

  I don’t want to step on anyone’s toes or take credit for the original work here. I’d just like to help get it into a shape where it can hopefully move forward upstream.

  If it’s useful, I’m happy to open a PR back to your branch or adjust things however you’d prefer.
  
  I also managed to fix the repeated keypress bug where only a single key would be sent to the client devices when holding a key down.

ZeroSegFault commented:
> @3l0w
> 
> I’ve been using this input-capture work for Deskflow for a while now and had a go at getting it working again on current Hyprland.
> 
> I’ve got both sides forward-ported and running locally now, and Deskflow is working for me.
> 
> Branches are here:
> 
> * Hyprland: https://github.com/ZeroSegFault/Hyprland/tree/zerosegfault/input-capture-hyprland
> * xdg-desktop-portal-hyprland: https://github.com/ZeroSegFault/xdg-desktop-portal-hyprland/tree/input-capture-portal
> 
> I don’t want to step on anyone’s toes or take credit for the original work here. I’d just like to help get it into a shape where it can hopefully move forward upstream.
> 
> If it’s useful, I’m happy to open a PR back to your branch or adjust things however you’d prefer.
> 
> I also managed to fix the repeated keypress bug where only a single key would be sent to the client devices when holding a key down.

I updated it again for the new the new updates on arch over the last few days.

Any one wanting to test try this

  ```
sudo pacman -S --needed git base-devel cmake ninja gcc pkgconf \
    wayland wayland-protocols hyprwayland-scanner \
    aquamarine hyprutils hyprlang hyprcursor hyprgraphics \
    libinput libxkbcommon libdrm mesa pixman cairo pango \
    libdisplay-info seatd libliftoff libxcb xcb-util-errors xcb-util-wm \
    xcb-util-renderutil xcb-util-cursor tomlplusplus \
    pipewire sdbus-cpp qt6-base qt6-wayland grim slurp
```

```
git clone -b zerosegfault/input-capture-hyprland https://github.com/ZeroSegFault/Hyprland
cd Hyprland
cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build -j$(nproc)
sudo cmake --install build

```
```
git clone -b input-capture-portal https://github.com/ZeroSegFault/xdg-desktop-portal-hyprland
cd xdg-desktop-portal-hyprland
cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build -j$(nproc)
sudo cmake --install build
```

Then restart Hyprland, restart the portal services, and test with Deskflow/Synergy using the Wayland/EI input path.

`systemctl --user restart xdg-desktop-portal-hyprland.service xdg-desktop-portal.service`

ayeganov commented:
This work is great - really looking forward to using it. Is this planned to be merged any time soon? I just switched to Omarchy from regolith-linux on Ubuntu and I really miss my synergy working with my mac M4.

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,563 @@
+#include <wayland-client-core.h>
+#include <wayland-util.h>
+
+namespace {
```
clang-format please

vaxerski commented on src/portals/InputCapture.cpp:
```diff
@@ -0,0 +1,563 @@
+#include <wayland-util.h>
+
+namespace {
+int32_t logicalSize(uint32_t size, double scale) {
```
prefer static methods rather than an anon ws.

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,73 @@
+#include <sdbus-c++/Types.h>
+
+enum ClientStatus {
+    CREATED,   //Is ready to be activated
```
too broad. CLIENT_STATUS_CREATED

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,73 @@
+#include "../shared/Session.hpp"
+#include <sdbus-c++/Types.h>
+
+enum ClientStatus {
```
: uint8_t

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,73 @@
+
+struct SBarrier {
+    uint id;
+    int  x1, y1, x2, y2;
```
no uninitialized primitives

vaxerski commented on src/portals/InputCapture.hpp:
```diff
@@ -0,0 +1,73 @@
+    std::unordered_map<std::string, const std::shared_ptr<SSession>> sessions;
+    //
+    std::unique_ptr<sdbus::IObject> m_pObject;
+    uint                            sessionCounter = 0;
```
all class (not struct) members are m_

`m_sessionCounter`

vaxerski reviewed (commented):
clang-format everything please
